cmake/version.cmake

```
@@ -21,3 +21,41 @@ add_custom_command(
 add_custom_target(ncs_version_h DEPENDS ${PROJECT_BINARY_DIR}/include/generated/ncs_version.h)
 
 add_dependencies(version_h ncs_version_h)
+
+if(${APP_VERSION_NUMBER})
+  add_custom_command(
+    OUTPUT ${PROJECT_BINARY_DIR}/include/generated/app_commit.h
+    COMMAND ${CMAKE_COMMAND} -DZEPHYR_BASE=${ZEPHYR_BASE} -DNRF_DIR=${NRF_DIR}
+      -DOUT_FILE=${PROJECT_BINARY_DIR}/include/generated/app_commit.h
+      -DCOMMIT_TYPE=APP
+      -DCOMMIT_PATH=${CMAKE_SOURCE_DIR}
+      -P ${ZEPHYR_NRF_MODULE_DIR}/cmake/gen_commit_h.cmake
+      DEPENDS ${CMAKE_SOURCE_DIR}/VERSION ${git_dependency}
+  )
+  add_custom_target(app_commit_h DEPENDS ${PROJECT_BINARY_DIR}/include/generated/app_commit.h)
+  add_dependencies(version_h app_commit_h)
+endif()
+
+add_custom_command(
+  OUTPUT ${PROJECT_BINARY_DIR}/include/generated/ncs_commit.h
+  COMMAND ${CMAKE_COMMAND} -DZEPHYR_BASE=${ZEPHYR_BASE} -DNRF_DIR=${NRF_DIR}
+    -DOUT_FILE=${PROJECT_BINARY_DIR}/include/generated/ncs_commit.h
+    -DCOMMIT_TYPE=NCS
+    -DCOMMIT_PATH=${NRF_DIR}
+    -P ${ZEPHYR_NRF_MODULE_DIR}/cmake/gen_commit_h.cmake
+    DEPENDS ${NRF_DIR}/VERSION ${git_dependency}
+)
+add_custom_target(ncs_commit_h DEPENDS ${PROJECT_BINARY_DIR}/include/generated/ncs_commit.h)
+add_dependencies(version_h ncs_commit_h)
+
+add_custom_command(
+  OUTPUT ${PROJECT_BINARY_DIR}/include/generated/zephyr_commit.h
+  COMMAND ${CMAKE_COMMAND} -DZEPHYR_BASE=${ZEPHYR_BASE} -DNRF_DIR=${NRF_DIR}
+    -DOUT_FILE=${PROJECT_BINARY_DIR}/include/generated/zephyr_commit.h
+    -DCOMMIT_TYPE=ZEPHYR
+    -DCOMMIT_PATH=${ZEPHYR_BASE}
+    -P ${ZEPHYR_NRF_MODULE_DIR}/cmake/gen_commit_h.cmake
+    DEPENDS ${ZEPHYR_BASE}/VERSION ${git_dependency}
+)
+add_custom_target(zephyr_commit_h DEPENDS ${PROJECT_BINARY_DIR}/include/generated/zephyr_commit.h)
+add_dependencies(version_h zephyr_commit_h)
```

A lot of duplication here.

```suggestion
set(types ncs zephyr)
set(types_upper NCS ZEPHYR)
set(path_vars NRF_DIR ZEPHYR_BASE)

if(${APP_VERSION_NUMBER})
  set(types ${types} app)
  set(types_upper ${types_upper} APP)
  set(path_vars ${path_vars} CMAKE_SOURCE_DIR)
endif()

foreach(type IN ZIP_LISTS types types_upper path_vars)
  add_custom_command(
    OUTPUT ${PROJECT_BINARY_DIR}/include/generated/${type_0}_commit.h
    COMMAND ${CMAKE_COMMAND} -DZEPHYR_BASE=${ZEPHYR_BASE} -DNRF_DIR=${NRF_DIR}
      -DOUT_FILE=${PROJECT_BINARY_DIR}/include/generated/${type_0}_commit.h
      -DCOMMIT_TYPE=${type_1}
      -DCOMMIT_PATH=${${type_2}}
      -P ${ZEPHYR_NRF_MODULE_DIR}/cmake/gen_commit_h.cmake
      DEPENDS ${${type_2}}/VERSION ${git_dependency}
  )
  add_custom_target(ncs_commit_h DEPENDS ${PROJECT_BINARY_DIR}/include/generated/ncs_commit.h)
  add_dependencies(version_h ${type_0}_commit_h)
endforeach()
```

Note, it's possible to use `foreach(type TUPPER path_var IN ZIP_LISTS types types_upper path_vars)` and then refer to those as `type` (was: type_0), `TUPPER` (was: type_1), `path_var` (was: type_2) in above code.